pkg/globalnet/controllers/node_controller.go

```diff
@@ -0,0 +1,220 @@
+			return nil, err
+		}
+	}
+
+	return controller, nil
+}
+
+func (n *nodeController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	node := from.(*corev1.Node)
+
+	// If the event corresponds to a different node which has globalIP annotation, release the globalIP back to Pool.
+	if node.Name != n.nodeName {
+		if existingGlobalIP := node.GetAnnotations()[SmGlobalIP]; existingGlobalIP != "" {
+			if op == syncer.Delete {
+				_ = n.pool.Release(existingGlobalIP)
+				return nil, false
+			} else {
+				_ = n.pool.Release(existingGlobalIP)
+				return n.updateNodeAnnotation(node, "")
+			}
+		}
+
+		return nil, false
+	}
+
+	cniIfaceIP := node.GetAnnotations()[constants.CNIInterfaceIP]
+	if cniIfaceIP == "" {
+		// To support connectivity from HostNetwork to remoteCluster, globalnet requires the
+		// cniIfaceIP of the respective node. Route-agent running on the node annotates the
+		// respective node with the cniIfaceIP. In this API, we check for the presence of this
+		// annotation and process the node event only when the annotation exists.
+		return nil, false
+	}
+
+	return n.onCreateOrUpdate(node)
+}
+
+func (n *nodeController) onCreateOrUpdate(node *corev1.Node) (runtime.Object, bool) {
+	klog.Infof("Node %q is updated", node.Name)
+
+	existingGlobalIP := node.GetAnnotations()[SmGlobalIP]
+	if existingGlobalIP != "" {
+		return nil, false
+	}
+
+	cniIfaceIP := node.GetAnnotations()[constants.CNIInterfaceIP]
+	globalIP, err := n.pool.Allocate(1)
+	if err != nil {
+		klog.Errorf("Error allocating IPs for node %q: %v", node.Name, err)
+		return nil, true
+	}
+
+	if err := n.iptIface.AddIngressRulesForHealthCheck(cniIfaceIP, globalIP[0]); err != nil {
+		klog.Errorf("Error programming rules for Gateway healthcheck on node %q: %v", node.Name, err)
+
+		_ = n.pool.Release(globalIP[0])
+
+		return nil, true
+	}
+
+	return n.updateNodeAnnotation(node, globalIP[0])
+}
+
+func (n *nodeController) reserveAllocatedIPs(federator federate.Federator, obj *unstructured.Unstructured) error {
+	existingGlobalIP := obj.GetAnnotations()[SmGlobalIP]
+	if existingGlobalIP == "" {
+		return nil
+	}
+
+	cniIfaceIP := obj.GetAnnotations()[constants.CNIInterfaceIP]
+	if cniIfaceIP == "" {
+		// To support Gateway healthCheck, globalnet requires the cniIfaceIP of the respective node.
+		// Route-agent running on the node annotates the respective node with the cniIfaceIP.
+		// In this API, we check for the presence of this annotation and process the node only
+		// when the annotation exists.
+		klog.Infof("cniIfaceIP annotation on node %q is currently missing", n.nodeName)
+		return nil
+	}
+
+	err := n.pool.Reserve(existingGlobalIP)
+	if err != nil {
+		klog.Errorf("Error reserving the existing globalIP to node %q, will re-allocate new one.", existingGlobalIP)
+	}
+
+	if err = n.iptIface.AddIngressRulesForHealthCheck(cniIfaceIP, existingGlobalIP); err != nil {
```

Comment:
We want to elide the IP table add if reserve failed.

```suggestion
	if err == nil {
		err = n.iptIface.AddIngressRulesForHealthCheck(cniIfaceIP, existingGlobalIP)
		if err != nil {
			_ = n.pool.Release(existingGlobalIP)
		}
	}
	
	if err != nil {
	    klog.Warningf("Could not reserve allocated GlobalIP for Node %q: %v", obj.GetName(), err)
```